dbt/dbt-run.py

```diff
@@ -133,10 +146,113 @@ def run_incremental_models():
+          if result is None or len(result) == 0:
+            return []
+          
+          start_timestamp, end_timestamp = result
+          start_timestamp = int(start_timestamp.timestamp())
+          end_timestamp = int(end_timestamp.timestamp())
+          batch_size = int(os.getenv("DBT_BATCH_SIZE") or 10000)
+          return [(start, min(start + batch_size, end_timestamp)) for start in range(start_timestamp, end_timestamp, batch_size)]
+
+def update_batch_status(timestamp, status):
+  with connection() as conn:
+    with conn.cursor() as cur:
+      # insert new entry
+      cur.execute(
+        f"INSERT INTO {os.getenv('POSTGRES_SCHEMA')}.dbt_batch_status (timestamp, status) VALUES (%s, %s);", [timestamp, status]
+      )
+      conn.commit()
+
+def get_last_processed_timestamp():
+  with connection() as conn:
+    with conn.cursor() as cur:
+      cur.execute(f"""
+          SELECT MAX(timestamp)
+          FROM {os.getenv('POSTGRES_SCHEMA')}.dbt_batch_status
+          WHERE status = 'success'
+      """)
+      result = cur.fetchone()
+      if result and result[0]:
+        return result[0]
+      return '1970-01-01 00:00:00'
+    
+def get_max_timestamp():
+  with connection() as conn:
+    with conn.cursor() as cur:
+      cur.execute(f"""
+          SELECT MAX(saved_timestamp)
+          FROM {os.getenv('POSTGRES_SCHEMA')}.document_metadata
+      """)
+      return cur.fetchone()[0]
+   
+def run_dbt_in_batches():
+  print("Running dbt in batches")
+  last_processed_timestamp = get_last_processed_timestamp()
+  batch_size = int(os.getenv("DBT_BATCH_SIZE") or 10000)
+  print(f"Starting new batch with timestamp: {last_processed_timestamp}")
+
+  while True:
+     print(f"Starting new batch with timestamp: {last_processed_timestamp}")
+     update_dbt_deps()
+     result = subprocess.run([
+        "dbt", "run",
+        "--profiles-dir", ".dbt",
+        "--vars", f'{{start_timestamp: "{last_processed_timestamp}", batch_size: {batch_size}}}'
+      ])
+
+     if result.returncode != 0:
+       print("Error running dbt")
+       update_batch_status(last_processed_timestamp, "error")
+       time.sleep(int(os.getenv("DATAEMON_INTERVAL") or 5))
+       continue
+     
+     update_batch_status(last_processed_timestamp, "success")
+     max_timestamp = get_max_timestamp()
+
+     if max_timestamp == last_processed_timestamp:
+       print("Finished processing all batches")
+       break
+     
+     last_processed_timestamp = max_timestamp
+
+  run_dbt()
+
+def run_dbt():
+  print("Starting regular dbt run")
+  while True:
+      update_models()
+      run_incremental_models()
+      time.sleep(int(os.getenv("DATAEMON_INTERVAL") or 5))
 
 if __name__ == "__main__":
+  print("Starting dbt run")
   setup()
-  while True:
-    update_models()
-    run_incremental_models()
-    time.sleep(int(os.getenv("DATAEMON_INTERVAL") or 5))
+  # check if we need to run in batch
+  pending_doc_count = get_pending_doc_count()
```

Comment:
For simplicity, I think we should just always run in batches. Otherwise, if we just `run_dbt` and then all the batching stuff timestamp indexes will become stale. 
If we have a period where dbt does not run or whatever (many docs pushed to db, externally), then records that have already been added to dbt will be added again.

Reply:
Yes, something like user transfers might trigger a lot of doc changes and rely on batch processing again to catch up.